hphp/runtime/vm/jit/relocation-arm.cpp

```diff
@@ -0,0 +1,951 @@
+ * so that the individual instructions in the sequence can be relocated
+ * one-by-one.
+ *
+ * NOTE: This helper will optimize sequences that are not smashables but which
+ *       are identical to smashables. Namely the sequences emitted for
+ *       jmp{} and jcc{} which use LDR literals for easy of patching at the
+ *       end of the translation. It's important to note that a smashable-like
+ *       sequence can be optimized during static relocation, but that an
+ *       actual smashable can't be optimized until live relocation and then
+ *       only if it has been smashed. Currently there is no means to know
+ *       if a smashable has been smashed, and so this optimization of
+ *       smashables during live relocation is not yet performed at all.
+ *
+ * dest_count, src_count and rewrites are updated to reflect when an
+ * instruction(s) is rewritten to a different instruction sequence.
+ */
+bool relocateSmashableHelper(RelocationInfo& rel, CGMeta& meta,
+                             CodeBlock& dest_block, TCA start, TCA end,
+                             TCA src, TCA dest, TCA jmp_dest,
+                             size_t& src_count, size_t& dest_count,
+                             const InstrSet& far, InstrSet& rewrites,
+                             bool& update_internal_refs) {
+
+  auto instr = Instruction::Cast(src);
+
+  /*
+   * A smashable jcc looks exactly like a smashable jmp with a preceding
+   * conditional branch. The checks below determine if there is a smashable
+   * jmp that can be optimized. It's important to not simply analyze this
+   * smashable jmp without first considering if it's part of a jcc.
+   */
+  auto target = smashableJmpTarget(src);
+  if (target) {
+    auto sl = getSmashableFromTargetAddr(src + kSmashJmpTargetOff);
+    if (sl && meta.smashableLocations.count(sl)) {
+      target = nullptr;
+    }
+  }
+
+  assertx(((uint64_t)target & 3) == 0);
+
+  if (target) {
+    auto adjusted = rel.adjustedAddressAfter(target);
+    if (!adjusted) adjusted = target;
+    int imm = (adjusted - dest) >> vixl::kInstructionSizeLog2;
+    if (is_int26(imm) && !far.count(instr)) {
+      vixl::MacroAssembler a { dest_block };
+      dest_block.setFrontier(dest);
+      a.b(imm);
+      src_count = smashableJmpLen() >> vixl::kInstructionSizeLog2;
+      dest_count = 1;
+      for (auto i = Instruction::Cast(src);
+           i < Instruction::Cast(src + src_count * kInstructionSize);
+           i = i->NextInstruction()) {
+        rewrites.insert(i);
+      }
+      update_internal_refs = true;
+      FTRACE(3,
+             "Relocated smashable at src 0x{:08x} with target 0x{:08x} to 0x{:08x} (0x{:08x})\n",
+             (uint64_t)src, (uint64_t)target, *((uint32_t*)dest), (uint64_t)dest);
+
+      return true;
+    }
+  }
+
+  /*
+   * Smashable calls are not yet supported, because we don't have any
+   * reloc debug info logged when smashing a call.
+   */
+
+  return false;
```

Comment:
Yep, that makes sense—I mostly just meant that you didn't need to leave an opening for handling calls, since you can just do it in a separate helper.